modules/javafx.graphics/src/main/java/com/sun/scenario/animation/AbstractPrimaryTimer.java

```diff
@@ -308,29 +317,80 @@ protected void timePulseImpl(long now) {
             debugNanos += fixedPulseLength;
             now = debugNanos;
         }
-        final PulseReceiver receiversSnapshot[] = receivers;
+
+        final ReceiverRecord<PulseReceiver>[] receiversSnapshot = receivers;
         final int rLength = receiversLength;
-        try {
-            receiversLocked = true;
-            for (int i = 0; i < rLength; i++) {
-                receiversSnapshot[i].timePulse(TickCalculation.fromNano(now));
+        receiversLocked = true;
+
+        for (int i = 0; i < rLength; i++) {
+            try {
+                receiversSnapshot[i].receiver().timePulse(TickCalculation.fromNano(now));
+            } catch (Throwable e) {
+                receiversSnapshot[i].handleException(e);
             }
-        } finally {
-            receiversLocked = false;
         }
+
+        receiversLocked = false;
         recordAnimationEnd();
 
-        final TimerReceiver animationTimersSnapshot[] = animationTimers;
+        final ReceiverRecord<TimerReceiver>[] animationTimersSnapshot = animationTimers;
         final int aTLength = animationTimersLength;
-        try {
-            animationTimersLocked = true;
-            // After every frame, call any frame jobs
-            for (int i = 0; i < aTLength; i++) {
-                animationTimersSnapshot[i].handle(now);
+        animationTimersLocked = true;
+
+        // After every frame, call any frame jobs
+        for (int i = 0; i < aTLength; i++) {
+            try {
+                animationTimersSnapshot[i].receiver().handle(now);
+            } catch (Throwable e) {
+                animationTimersSnapshot[i].handleException(e);
             }
-        } finally {
-            animationTimersLocked = false;
         }
+
+        animationTimersLocked = false;
     }
 
+    private static abstract class ReceiverRecord<T> {
```

Comment:
All this does look complicated, just to limit the number of messages in the log (a noble cause).

What if we just print the first message and ignore the rest?  It might be relatively unlikely that two different failures appeared at the same time, each with its own root cause?

If we just show the first one, we'll remove all this complexity and the system property, and possibly get back to the simpler original code?

What do you think?

Reply:
We're throwing away all exceptions after a certain point (either after the first, or some number of exceptions after that). A potential use case would be debugging a large application that for some reason throws lots of different exceptions from timers. There can't be many applications that do this, because right now that would most likely just freeze the application. I'm okay either way (logging the first exception, or logging the first 100 exceptions with a user-configurable threshold).

Reply:
We could also just keep it with a fixed threshold, but remove the system property.

Reply:
It just seems unnecessary to add all this complexity.  The first exception is all we need, really - we can skip the 2nd, 3rd, ...  This might rule out the property.

It might be unlikely that two or more completely unrelated failures would happen in the field in a well-tested application [citation needed], I think there is a little value in trying to add this complex logic per timer.  A static boolean might suffice.

Reply:
On the other hand, AnimationTimer is a public API that allows for arbitrary code execution, which always has the potential to fail for any number of reasons (think of calling an unstable API from the timer callback). We can’t rule that out entirely. This is only a fallback in any case, as a well-tested application shouldn’t throw exceptions from timer callbacks in the first place.

Reply:
I've removed the system property, so there's a hard-coded limit of 100 exceptions now.

Reply:
100 still looks excessive to me - it's unlikely that we'll have two different scenarios when this happens, so most of the time the log will have 100 identical traces.

I mean, one is probably enough, we could have 2 or 4 just to drive the message home.  Anything beyond 8 is simply annoying, don't you think?

Reply:
In general, we shouldn't suppress any exceptions that would otherwise be logged. The only thing that makes this case a bit different is that with a periodic timer, there's the potential for an inflated log that grows without bounds. I think that defining a reasonable upper limit is enough to prevent that; in particular, I think we shouldn't try to assume that any particular number of logged failures (like 1 or 4) is "enough" for application developers.

Reply:
well, I still think it's way too high of a number, but the expectation is that once the dev sees a 100 traces in the log they'll fix it right away.

And if some company finds that all of a sudden their client logs exploded in size, they should have invested more in QA.